### ISSUE TITLE
Backblaze setup example to direct upload (CORS)

### ISSUE THREAD

**Issue by antoniosmgatto** [jeygeethan/activestorage-backblaze#4]
Hi There !!!

First of all, Thanks for this integration !

I'm facing some CORS trouble with B2 and I appreciate any help.

I can't setup my bucket to upload files from my frontend (React-app). I am using this gem with Active Storage to generate the direct upload url. 

This is the command I use to setup the CORS rules:

```bash
b2 update-bucket --corsRules '[
    {
        "corsRuleName": "downloadFromAnyOriginWithUpload",
        "allowedOrigins": [
            "*"
        ],
        "allowedHeaders": [
            "*"
        ],
        "allowedOperations": [
            "b2_download_file_by_id",
            "b2_download_file_by_name",
            "b2_upload_file",
            "b2_upload_part"
        ],
        "exposeHeaders": [
            "authorization",
            "x-bz-file-name",
            "x-bz-part-number",
            "x-bz-content-sha1"
        ],
        "maxAgeSeconds": 3600
    }
]' my-bucket allPublic
```

Any idea?

Thanks

**Comment by jeygeethan** [comment]
Hey Can you let me know what is the error that you are getting on the frontend react app? Are you getting the upload url in the first place?

**Comment by antoniosmgatto** [comment]
Yes, I get the url from my backend.

console tab output

![image](https://user-images.githubusercontent.com/5853930/98405373-ee24f400-204a-11eb-80c2-7b43d70fdd9c.png)

upload request

![image](https://user-images.githubusercontent.com/5853930/98405475-1ad90b80-204b-11eb-81cc-383ade987fac.png)

pre-flight request 

![image](https://user-images.githubusercontent.com/5853930/98405541-2f1d0880-204b-11eb-9e9e-114c372a4119.png)
![image](https://user-images.githubusercontent.com/5853930/98405549-32b08f80-204b-11eb-80c7-74a4e8e7db48.png)

Any idea?

**Comment by jeygeethan** [comment]
Hi,

from the first screenshot, it looks like CORs is not enabled. I usually set the cors for the bucket in the UI.

Can you try to do that?

**Comment by jeygeethan** [comment]
Try this out - https://www.backblaze.com/blog/enable-cors-for-cross-origin-resource-sharing/#:~:text=Backblaze%20supports%20the%20standard%20CORS,of%20assets%20to%20other%20origins.

**Comment by antoniosmgatto** [comment]
Hi there !!

The cors is enabled.  Here my settings:

```bash
[
  {
    "corsRuleName": "downloadFromAnyOriginWithUpload",
    "allowedOrigins": [
      "*"
    ],
    "allowedHeaders": [
      "*"
    ],
    "allowedOperations": [
      "b2_download_file_by_id",
      "b2_download_file_by_name",
      "b2_upload_file",
      "b2_upload_part"
    ],
    "exposeHeaders": [
      "authorization",
      "content-type",
      "content-length",
      "x-bz-file-name",
      "x-bz-part-number",
      "x-bz-content-sha1"
    ],
    "maxAgeSeconds": 3600
  }
]
```

I tried to upload a file with ruby and it worked. Have you ever used direct upload?.

**Comment by EmCousin** [comment]
Hi there,

I have been encountering the exact same issue, and after digging for a while I managed to perform a direct upload to B2 successfully, but via a `POST` HTTP request. Performing a `PUT` request, as [ActiveStorage does by default](https://github.com/rails/rails/blob/main/activestorage/app/javascript/activestorage/blob_upload.js#L9) however still fails with the same `CORS method not allowed` error message.

It appears BackBlaze does not enable PUT methods when using their B2 API. I have reached customer support and am waiting for them to answer.

**Comment by EmCousin** [comment]
Here is the answer from Backblaze Support:

> After further observation, it looks like you're using the B2 API for PUT actions rather than the S3 API. Essentially, the B2 API accepts POST calls but not PUT calls while the S3 API accepts PUT calls but not POST calls. So in this case, the CORS rule wouldn't be the cause here.From here, please try using the S3 API to conduct the PUT call instead. Please let me know if you have any further questions or concerns.

If this gem is going to be an ActiveStorage adapter for Backblaze, it could try to use the S3 API instead

**Comment by jeygeethan** [comment]
HI @EmCousin, This gem was created before b2 had the s3 apis. There was no other gem to have it integrated with the activestorage. 

Considering that they have the s3 apis, i am guessing you can use the aws-sdk gem itself and just changing the s3 endpoint to b2. I haven't done this, but just guessing. If you can help with the results it would be great.

**Comment by EmCousin** [comment]
Hey @jeygeethan, I actually did that yesterday and indeed using the aws-sdk gem works just fine. With Rails and Active Storage, the following configuration suffices:

```yaml
backblaze:
  service: S3
  access_key_id: <KEY_ID>
  secret_access_key: <KEY_TOKEN>
  region: <REGION> # example: us-east-001
  bucket: <BUCKET_NAME>
  endpoint: https://s3.<REGION>.backblazeb2.com
``` 

Providing the `endpoint` is necessary as it otherwise defaults to AWS'

Still, good job on making a gem to wire BackBlaze with ActiveStorage using the B2 API. If only we could configure Active Storage direct upload requests' verb to be POST by default, direct upload would work like a charm.

**Comment by jeygeethan** [comment]
@EmCousin I will try my best to see if I can patch the direct uploads verb to POST. But I am curious, will you be using the b2 apis instead of the s3?

**Comment by EmCousin** [comment]
>  I will try my best to see if I can patch the direct uploads verb to POST

@jeygeethan Unfortunately you can't. The direct upload request's verb is hardcoded in [Active Storage's JS module](https://github.com/rails/rails/blob/main/activestorage/app/javascript/activestorage/blob_upload.js#L9)

The S3-compatible API suits my needs, but using the B2 APi would be much better to enjoy the full set of features Backblaze provides. But that is out of the scope of this issue

**Comment by jeygeethan** [comment]
@EmCousin I forgot to add more one thing in the readme. 

I had actually extracted the activestorage-backblaze and actiontext-backblaze npm packages separately but forgot to put them on github. Here they are:

[https://github.com/jeygeethan/actiontext-backblaze](https://github.com/jeygeethan/actiontext-backblaze)
[https://github.com/jeygeethan/activestorage-backblaze-javascript](https://github.com/jeygeethan/activestorage-backblaze-javascript)

If you see, it has been updated to POST at https://github.com/jeygeethan/activestorage-backblaze-javascript/blob/master/src/blob_upload.js#L9

Add this to your package.json

`"@jeygeethan/actiontext-backblaze": "^6.1.3-alpha"
`

**Comment by tbhockey** [comment]
> Hey @jeygeethan, I actually did that yesterday and indeed using the aws-sdk gem works just fine. With Rails and Active Storage, the following configuration suffices:
> 
> ```yaml
> backblaze:
>   service: S3
>   access_key_id: <KEY_ID>
>   secret_access_key: <KEY_TOKEN>
>   region: <REGION> # example: us-east-001
>   bucket: <BUCKET_NAME>
>   endpoint: https://s3.<REGION>.backblazeb2.com
> ```
> 
> Providing the `endpoint` is necessary as it otherwise defaults to AWS'
> 
> Still, good job on making a gem to wire BackBlaze with ActiveStorage using the B2 API. If only we could configure Active Storage direct upload requests' verb to be POST by default, direct upload would work like a charm.

Broooo thank you so so much. Not sure why this reply doesn't have more love, but it deserves an award. This had been giving our whole team headaches for days.

**Comment by jeygeethan** [comment]
@tbhockey True, since this was created before backblaze support S3 apis, this gem was relevant back then. But now you can directly use the S3 endpoints and skip this gem. I should probably add this in the readme!

**Comment by tbhockey** [comment]
@jeygeethan that would probably be very helpful for others. tbh this gem was perfectly fine for us, it was just until trying to use this DirectUpload feature of Rails that we ran into this issue as the regular BB endpoints don't support it. But thankfully their S3 apis do and this works beautifully.

**Comment by adenta** [comment]
Dang, I'm using the `@jeygeethan/activestorage-backblaze` npm package, _and_ backblaze is configured for `Share everything in this bucket with every origin`, and I _still_ can't get CORS requests to pass. Do I need to RTFM or are there any other tricks I can try?
